File: sparkcql-core/src/main/scala/gov/va/sparkcql/adapter/model/ModelAdapterAggregator.scala
package gov.va.sparkcql.adapter.model

import scala.reflect.runtime.universe._
import org.apache.spark.sql.types.StructType
import javax.xml.namespace.QName
import gov.va.sparkcql.io.Log
import javax.naming.OperationNotSupportedException

sealed class ModelAdapterAggregator(modelAdapters: List[ModelAdapter]) extends ModelAdapter {

  lazy val namespaceUri: String = throw new OperationNotSupportedException()

  lazy val supportedDataTypes: List[QName] = {
    modelAdapters.flatMap(_.supportedDataTypes).distinct
  }

  def resolveModel(dataType: QName): Option[ModelAdapter] = {
    modelAdapters.filter(_.supportedDataTypes.contains(dataType)).headOption
  }
  
  def ensureDataTypeSupported(dataType: QName): Unit = {
    if (supportedDataTypes.contains(dataType)) {
      throw new UnsupportedOperationException(String.format("No model adapter supports this operation for type '%s'.", dataType.toString()));
    }
  }

  def schemaOf(dataType: QName): Option[StructType] = {
    val resolved = resolveModel(dataType)
    if (resolved.isDefined) {
      resolved.get.schemaOf(dataType)
    } else {
      Log.warn(s"Unable to find schema ${dataType.toString()}")
      None
    } 
  }

  def metaInterval(typeName: String): (String, String) = ("low", "high")    // TODO: Throw error or check all

  def typeToElmMapping(typeName: String): Map[String, String] = throw new OperationNotSupportedException()
}

File: sparkcql-core/src/main/java/gov/va/sparkcql/adapter/model/ModelAdapterAggregator.java
package gov.va.sparkcql.adapter.model;

import java.util.List;

import javax.xml.namespace.QName;

import org.apache.spark.sql.types.StructType;

public class ModelAdapterAggregator implements ModelAdapter {

    protected List<ModelAdapter> modelAdapters;

    public ModelAdapterAggregator(List<ModelAdapter> modelAdapters) {
        this.modelAdapters = modelAdapters;
    }

    ModelAdapter resolveModel(QName dataType) {
        var resolvedAdapter = modelAdapters.stream().filter(a -> a.supportedDataTypes().contains(dataType));
        if (resolvedAdapter.count() > 0) {
            return resolvedAdapter.findFirst().get();
        } else {
            throw new RuntimeException("No ModelAdapter found for data type " + dataType.toString() + ".");
        }
    }

    @Override
    public String namespaceUri() {
        throw new UnsupportedOperationException();
    }

    @Override
    public List<QName> supportedDataTypes() {
        return this.modelAdapters.stream().flatMap(adapter -> {
            return adapter.supportedDataTypes().stream();
        }).distinct().toList();
    }

    @Override
    public StructType schemaOf(QName dataType) {
        return resolveModel(dataType).schemaOf(dataType);
    }    
}
